component/module/module.go

```diff
@@ -16,20 +16,34 @@ type ModuleComponent struct {
 	opts component.Options
 	mod  component.Module
 
-	mut           sync.RWMutex
-	health        component.Health
-	latestContent string
-	latestArgs    map[string]any
+	mut                         sync.RWMutex
+	health                      component.Health
+	latestContent               string
+	latestArgs                  map[string]any
+	latestParentDeclareContents map[string]string
 }
 
 // Exports holds values which are exported from the run module.
+// This export type is deprecated.
 type Exports struct {
 	// Exports exported from the running module.
 	Exports map[string]any `river:"exports,block"`
 }
 
 // NewModuleComponent initializes a new ModuleComponent.
 func NewModuleComponent(o component.Options) (*ModuleComponent, error) {
+	c := &ModuleComponent{
+		opts: o,
+	}
+	var err error
+	c.mod, err = o.ModuleController.NewModule("", func(exports map[string]any) {
+		c.opts.OnStateChange(exports)
+	})
+	return c, err
+}
+
+// TODO: Remove when getting rid of old modules
+func NewModuleComponentDeprecated(o component.Options) (*ModuleComponent, error) {
```

Comment:
I'm not sure what the difference between `NewModuleComponent/NewModuleComponentDeprecated` is, can we expand on this?


To minimize breaking changes to the API, the Go convention is to mark the current function with a deprecation notice in the docstring and add the new one under a different name. We're still pre v1.0, but it'd be a good exercise to figure out how you'd handle this instead.

Reply:
I added the depretation notice and more doc

Reply:
Do we _need_ two separate functions here? They look identical to.

Reply:
The callback passed to the NewModule is different: in the deprecated version it passes the Exports type while in the new version it passes the map[string]any. As a result, when using new modules you don't need to write ".exports" in you config. It makes it less verbose and it acts like the other components.
I decided to use two functions because:
- it does not touch the old logic which reduces the risk of regression for the old modules when we merge this
- when we will get rid of the old modules, we can just remove the deprecated part without having to touch the new modules logic (less risk of regression for the new modules)
But I'm also open to suggestions if you have another solution in mind